stardis/plasma/molecules.py

```diff
@@ -0,0 +1,421 @@
+            ion2_number_density = ion_number_density.loc[
+                molecule_row.Ion2, molecule_row.Ion2_charge
+            ]
+
+            # Barklem and Collet 2016 equilibrium constants are pressure constants and are in SI units
+            pressure_equilibirium_const_at_depth_point = np.interp(
+                t_electrons,
+                equilibrium_const_temps,
+                atomic_data.molecule_data.equilibrium_constants.loc[molecule].values,
+            )
+            # Convert from pressure constants to number density constants using ideal gas law
+            equilibirium_const_at_depth_point = (
+                (
+                    10 ** (pressure_equilibirium_const_at_depth_point)
+                    * (u.N / u.m**2)
+                    / (const.k_B * t_electrons * u.K)
+                )
+                .to(u.cm**-3)
+                .value
+            )
+
+            molecule_number_density = (ion1_number_density * ion2_number_density) / (
+                equilibirium_const_at_depth_point
+            )
+
+            number_densities_arr[
+                atomic_data.molecule_data.equilibrium_constants.index.get_loc(molecule)
+            ] = molecule_number_density
+
+        molecule_densities_df = pd.DataFrame(
+            number_densities_arr,
+            index=atomic_data.molecule_data.equilibrium_constants.index,
+            columns=ion_number_density.columns,
+        )
+        # Keep track of the individual ions - useful to calculate molecular masses later for doppler broadening
+        molecule_ion_map = pd.DataFrame(
+            molecules_df[["Ion1", "Ion2"]],
+        )
+
+        return molecule_densities_df, molecule_ion_map
+
+    def preprocess_ion(self, molecules_df, ion):
+        """
+        Preprocesses a component ion in the molecule data to add the ion's atomic number and charge to the df.
+        """
+        molecules_df[
+            [f"Ion{ion}_symbol", f"Ion{ion}_positive", f"Ion{ion}_negative"]
+        ] = molecules_df[f"Ion{ion}"].str.extract(r"([A-Z][a-z]?+)(\+*)(\-*)")
+        molecules_df[f"Ion{ion}"] = molecules_df[f"Ion{ion}_symbol"].apply(
+            element_symbol2atomic_number
+        )
+        molecules_df[f"Ion{ion}_charge"] = molecules_df[f"Ion{ion}_positive"].apply(
+            len
+        ) - molecules_df[f"Ion{ion}_negative"].apply(len)
+        return
+
+
+class MoleculePartitionFunction(ProcessingPlasmaProperty):
+    """
+    Processes the partition function for each molecule at each depth point by interpolating the partition function data.
+    From Barklem and Collet 2016.
+    Attributes
+    ----------
+    molecule_partition_function : DataFrame
+            A pandas DataFrame with dtype float. This represents the partition function
+            for each molecule at each depth point.
+    """
+
+    outputs = ("molecule_partition_function",)
+
+    def calculate(self, t_electrons, atomic_data):
+        partition_functions = pd.DataFrame(
```

Comment:
Probably should not initialize dataframe like that - make an array of the data and build dataframe from that

Reply:
I've realized I did this because there's not a good way to vectorize this input. Each row needs a new interpolated function because the partition functions for each molecule will be different. I think you'd still need to do an iterrows or apply which I don't think are particularly better.

Also, the dataframe being created is only about ~300 long so the performance isn't horribly painful.